Get the editor from git through `git var`

mathijs81 opened the pull request:
We were invoking git anyway for the git config value of core.editor so we can just as well let git return the result of its full logic instead and simplify our code. This also handles the case where e.g. debian/ubuntu have compiled git with a different fallback than the "vi" default.
I kept the GIT_EDITOR check to be able to skip the git invocation if that's handy, and have the extremely simple vi/notepad fallback at the end still just in case something goes wrong with git.

Byron commented:
Thanks a lot, I think that's *the* way to do it.
With the refactorings we discussed on Discord I think this is basically ready to go - please feel free to request a review when ready.

mathijs81 committed:
Get the editor from git through `git var`

We were invoking git anyway for the git config value of core.editor so
we can just as well let git return the result of its full logic
instead. This also handles the case where e.g. debian/ubuntu have
compiled git with a different fallback than the "vi" default.

mathijs81 committed:
Spawn editor using shell like git does

This fixes e.g. "code --wait" as the editor setting.

mathijs81 commented on crates/but/src/tui/get_text.rs:
```diff
@@ -87,139 +76,35 @@ where
+            actual, "from-GIT_EDITOR",
+            "GIT_EDITOR should take precedence if git is executed correctly"
         );
     }
```
Intended, we don't want to write tests for git itself

Byron committed:
refactor

- be more verbose when Git couldn't be invoked to ease remote debugging later.
- don't force vars to strings, retain their original encoding
- avoid hashmap in favor of iterator

Byron committed:
Avoid manual tempfile implementation in `get_text::from_editor()`

Let's set an example in code that is supposed to be used in many comands,
more akind to a CLI-STD.

Byron committed:
Use `BString` as editor return value to not enforce encodings.

This will make it harder to call from legacy code, but will make it
clear for modern code that we don't know the encoding and shouldn't
lighthartedly butcher it by converting to unicode lossily.

Note that all callers now state explicitly that they handle the message
lossily, and we should fix this once the code is modernised.

While at it, also avoid rewriting line endings in favor of reusing
the ones that the user specified.

Byron committed:
Address copilot review

Byron commented on crates/but/src/tui/get_text.rs:
```diff
@@ -1,225 +1,124 @@
+    let filtered_lines: Vec<&BStr> = content
+        .lines_with_terminator()
+        .filter(|line| !line.trim_start().starts_with_str("#"))
+        .map(|line| line.as_bstr())
```
Not true, unfortunately.

Byron commented on crates/but/src/tui/get_text.rs:
```diff
@@ -1,225 +1,124 @@
+/// Note that this string must be valid in filenames.
+///
+/// Returns the edited text (*without known encoding*) verbatim.
+pub fn from_editor(identifier: &str, initial_text: &str) -> Result<BString> {
```
Great catch

Byron reviewed (approved):
Thanks a lot! I particularly like all the removed tests which felt a bit like an exercise before, and of course, the overall simplification.

While at it, I made the code a bit more idiomatic and also looked at the public functions, which I didn't do before. From there, I put it into a shape that I think should be suitable for a Cli-std if you will.

All my changes have been validated with `cargo run --bin but  -- -C <repo-path> describe <short-hash>` (only).

Please do take a look and give it another refactoring round if you see something to improve.

Byron committed:
Fix cargo clippy issues that show up with Rust 1.92

Byron committed:
Switch to Rust 1.92 officially (as our image seems to use it, and CI will follow soon)
